### PR TITLE
DBUP Integration

### PR DESCRIPTION
**Pull request by marwanrahman**
This PR is to replace old code that is no longer necessary with DBUP functions. The main changes include:

- Dataset creation from a SQL query using view_dataset_creator.py
- Artifact logging
- Simplified training script

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -1,456 +1,264 @@
...
-    class_weight = get_class_weights(histogram)
+    :return: keras model
+    """
 
     # Define performance metrics
     classes = cfg['DATA']['CLASSES']
-    # F1Score(name='f1score', num_classes=2)
+    threshold = 1.0 / len(classes)  # Binary classification threshold for a class
+
     metrics = ['accuracy', AUC(name='auc'), Precision(name='precision', thresholds=0.5),
```
Comment (delaneysmith22) [inline src/train.py:111]: Should we include Specificity as a metric here?

**Review by delaneysmith22 — COMMENTED** [review]
Great job! As an aside, I'm wondering if we should also integrate DBUP's visualization module in this PR once Saad has finished it up (depending on the timeline).

File: src/models/models.py [hunk 2]
```diff
@@ -101,13 +103,15 @@ def mobilenetv2(model_config, input_shape, metrics, n_classes, output_bias=None)
     model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=metrics)
     return model
 
-def inception_v3(model_config, input_shape, metrics, n_classes, output_bias=None):
+
+def inception_v3(model_config, input_shape, metrics, n_classes, output_bias=None, weights_path=None):
```
[the remark this answers is not in the document]
Comment (farazali7) [reply src/models/models.py:107]: Good idea! We already have a models module in DBUP too. One thing to be careful of if we move ahead with doing this is only modularizing the base parts of each of the models since the "head" layers and their parameters are usually always project-specific.

File: src/train.py [hunk 3]
```diff
@@ -1,456 +1,264 @@
...
     Computes weights for each class to be applied in the loss function during training.
-    :param histogram: A list depicting the number of each item in different class
-    :param class_multiplier: List of values to multiply the calculated class weights by. For further control of class weighting.
+    :param train_df: training DataFrame
     :return: A dictionary containing weights for each class
-    '''
+    """
+
+    # Get class distribution
+    histogram = np.bincount(train_df['Class'].to_numpy().astype(int))
```
Comment (farazali7) [inline src/train.py:31]: Could make the key configurable in case someone has a differently formatted df.

File: src/train.py [hunk 4]
```diff
@@ -1,456 +1,264 @@
...
-def define_callbacks(cfg):
-    '''
+def compute_output_bias(train_df: pd.DataFrame) -> Constant:
+    """
+    Determines the bias on the model final layer
+    :param train_df: training DataFrame
+    :return: initializer that has constant values corresponding to the bias value
+    """
+    # Compute output bias
+    histogram = np.bincount(train_df['Class'].astype(int))
```
Comment (farazali7) [inline src/train.py:60]: Same key comment as above. In fact I think we can also make a separate function to compute the histogram since it's being computed in both class weight and output bias computation functions.

**Review by farazali7 — COMMENTED** [review]
Nice work Marwan - this really makes the repo look a lot cleaner. Left some minor comments for review.

File: README.md [hunk 5]
```diff
@@ -58,25 +58,46 @@ Waterloo, Canada). Following this, all ultrasound clips were deconstructed into
...
+    SSH_USERNAME: "azureuser"
+    PRIVATE_KEY: (path to private key)
+    DB_HOST: (database host name)
+    DB_PORT: 3306
+    LOCAL_BIND_IP: "127.0.0.1"
+    DB_USERNAME: (database user name)
+    DB_PASSWORD: (database password)
+    DATABASE_NAME: data
+    ```
+3. Copy auto_masking_deeper.h5 to the data/ directory. Or add the _keypoint_model_ folder to the data/ directory.
```
[the remark this answers is not in the document]
Comment (bvanber) [reply README.md:80]: I almost wonder if we want this readme to be in our confluence library and not in git... We can be more vauge for our public viewers

File: README.md [hunk 6]
```diff
@@ -58,25 +58,46 @@ Waterloo, Canada). Following this, all ultrasound clips were deconstructed into
...
    
 ## Use Cases
 
 ### Train a Model
 
 With a pre-processed clip dataset, you can train a frame classification model of a chosen model definition.
 1. Assemble in a pre-processed clip dataset (see [**_Building a Dataset_**](#building-a-dataset)) and set the appropriate data paths in [_config.yml_](config.yml). 
-2. Mask the images of extraneous information outside the ultrasound beam. In our group, this was done with proprietary software mentioned above in 'Building a Dataset'
-3. Generate a frame dataset from the masked clips using [_build-dataset.py_](/src/data/build-dataset.py).
 4. Set the desired data and train configuration fields in [_config.yml_](config.yml) including setting a model type using the `MODEL_DEF` parameter and setting the `EXPERIMENT_TYPE` to _single_train_.
```
Comment (bvanber) [inline README.md:101]: If we are keeping these steps can we update the list numbers

File: src/models/models.py [hunk 7]
```diff
@@ -17,6 +17,8 @@
 from tensorflow.keras.initializers import Constant
 
 import tensorflow as tf
+from src.models.model_utils import *
```
Comment (bvanber) [inline src/models/models.py:20]: Can we please remove the * import (even though it is our own package lol, just god practice)

File: src/train.py [hunk 8]
```diff
@@ -1,456 +1,264 @@
...
-    # Set up preprocessing transformations to apply to each item in dataset
-    preprocessor = Preprocessor(preprocessing_fn)
-    train_set = preprocessor.prepare(train_set, shuffle=True, augment=True)
-    val_set = preprocessor.prepare(val_set, shuffle=False, augment=False)
-    test_set = preprocessor.prepare(test_set, shuffle=False, augment=False)
-
-    # Get class weights based on prevalences
-    histogram = np.bincount(train_df['Class'].to_numpy().astype(int))  # Get class distribution
-    class_weight = get_class_weights(histogram)
+    :return: keras model
```
Comment (bvanber) [inline src/train.py:104]: Can we add the exact data type here

File: src/train.py [hunk 9]
```diff
@@ -1,456 +1,264 @@
...
+
+    # Optionally get path to pretrained weights
+    pretrained_path = cfg['PATHS']['PRETRAINED_WEIGHTS'] if cfg['TRAIN']['USE_PRETRAINED'] else None
 
+    # Train the model
+    model = train_model(model_def, train_set, val_set, val_df, hparams, output_bias=output_bias,
+                        class_weight=class_weight, pretrained_path=pretrained_path)
+
+    # Save the model's weights
+    if experiment != 'cross_validation' and cfg['TRAIN']['SAVE_MODEL']:
```
Comment (bvanber) [inline src/train.py:220]: May be nice to have the option to save cross val model weights. As unlikely as I think it is that we will need that

File: config.yml [hunk 10]
```diff
@@ -19,21 +19,74 @@ PATHS:
...
+    CLIPS_JOIN: id
+    FRAMES_JOIN: id
+  ARTIFACT_SEED: 45
 
 DATA:
   IMG_DIM: [128, 128]
   VAL_SPLIT: 0.1
   TEST_SPLIT: 0.1
   CLASSES: ['parenchymal', 'pleural']   # Classes for binary classification
+  MASKING:
```
Comment (bvanber) [inline config.yml:61]: Is info about our masking mthod public? Do we want to make it public? Also this could be a good spot to add documentation for these params (and any other that are not clearly evident by their name)
